examples/sparse_regression.py

```diff
@@ -0,0 +1,318 @@
+    quadratic_weights = torch.tensor(quadratic_weights)
+
+    # we standardize the outputs Y
+    Y = Y_mean
+    Y -= Y.mean()
+    Y_std1 = Y.std()
+    Y /= Y_std1
+    Y += sigma_obs * torch.randn(N)
+    Y -= Y.mean()
+    Y_std2 = Y.std()
+    Y /= Y_std2
+
+    assert X.shape == (N, P)
+    assert Y.shape == (N,)
+
+    return X, Y, singleton_weights / (Y_std1 * Y_std2), expected_quad_dims
+
+
+def main(args):
+    # setup hyperparameters for the model
+    hypers = {'expected_sparsity': max(1.0, args.num_dimensions / 10),
+              'alpha1': 3.0, 'beta1': 1.0, 'alpha2': 3.0, 'beta2': 1.0, 'alpha3': 1.0,
+              'c': 1.0, 'alpha_obs': 3.0, 'beta_obs': 1.0}
+
+    P = args.num_dimensions
+    S = args.active_dimensions
+    Q = args.quadratic_dimensions
+
+    # generate artificial dataset
+    X, Y, expected_thetas, expected_quad_dims = get_data(N=args.num_data, P=P, S=S,
+                                                         Q=Q, sigma_obs=args.sigma)
+
+    loss_fn = Trace_ELBO().differentiable_loss
+
+    # We initialize the AutoDelta guide (for MAP estimation) with args.num_trials many
+    # initial parameters sampled from the vicinity of the median of the prior distribution
+    # and then continue optimizing with the best performing initialization.
+    init_losses = []
+    for restart in range(args.num_restarts):
+        pyro.clear_param_store()
+        pyro.set_rng_seed(restart)
+        guide = AutoDelta(model, init_loc_fn=functools.partial(init_to_median, num_samples=50))
+        with torch.no_grad():
+            init_losses.append(loss_fn(model, guide, X, Y, hypers).item())
+
+    pyro.set_rng_seed(np.argmin(init_losses))
+    pyro.clear_param_store()
+    guide = AutoDelta(model, init_loc_fn=functools.partial(init_to_median, num_samples=50))
+    with torch.no_grad():
+        loss_fn(model, guide, X, Y, hypers)
+    # we also make sure the initial observation noise is not too large
+    # (otherwise we run the danger of getting stuck in bad local optima during optimization).
+    pyro.param('auto_var_obs').unconstrained().data -= 2.0
+
+    with poutine.block(), poutine.trace(param_only=True) as param_capture:
+        guide(X, Y, hypers)
+
+    # Instead of using pyro.infer.SVI and pyro.optim we instead construct our own PyTorch
+    # optimizer and take charge of gradient-based optimization ourselves.
+    params = list([pyro.param(name).unconstrained() for name in param_capture.trace])
+    adam = Adam(params, lr=args.lr)
+
+    report_frequency = 50
+    print("Beginning MAP optimization...")
+
+    # the optimization loop
+    for step in range(args.num_steps):
+        loss = loss_fn(model, guide, X, Y, hypers) / args.num_data
+        loss.backward()
+        adam.step()
+        adam.zero_grad()
+
+        # we manually reduce the learning rate according to this schedule
+        if step in [100, 300, 700, 900]:
+            adam.param_groups[0]['lr'] *= 0.2
```

Comment:
there's a [scheduler](https://pytorch.org/docs/stable/optim.html?highlight=scheduler#torch.optim.lr_scheduler.MultiStepLR) for this

Reply:
I wanted to control everything myself and show that's done